lib/marc/datafield.rb

```diff
@@ -123,52 +141,34 @@ def append(subfield)
       @subfields.push(subfield)
     end
 
-    
-
-    # You can iterate through the subfields in a Field:
-    #   field.each {|s| print s}
-
-    def each
-      for subfield in subfields
-        yield subfield
-      end
-    end
-
-    #def each_by_code(filter)
-    #  @subfields.each_by_code(filter)
-    #end
-
-    # You can lookup subfields with this shorthand. Note it 
+    # You can lookup subfields with this shorthand. Note it
     # will return a string and not a MARC::Subfield object.
+    #
+    # NOTE: the value of only the *first* subfield is returned
     #   subfield = field['a']
-    
+
     def [](code)
-      subfield = self.find {|s| s.code == code}
+      subfield = self.detect {|s| s.code == code}
       return subfield.value if subfield
-      return
+      return nil
     end
- 
+
 
     def codes(dedup=true)
-      codes = []
-      @subfields.each {|s| codes << s.code }
-      dedup ? codes.uniq : codes
+      codes = subfields.map(&:code)
+      codes.uniq! if dedup
+      codes
     end
 
-    # Two fields are equal if their tag, indicators and 
+    # Two fields are equal if their tag, indicators and
     # subfields are all equal.
 
     def ==(other)
-      if @tag != other.tag
-        return false 
-      elsif @indicator1 != other.indicator1
-        return false 
-      elsif @indicator2 != other.indicator2
-        return false 
-      elsif @subfields != other.subfields
-        return false
-      end
-      return true
+      tag == other.tag and
+        indicator1 == other.indicator1 and
+        indicator2 == other.indicator2 and
+        subfields.size == other.subfields.size and
+        subfields.zip(other.subfields).all?{|a| a[0] == a[1] }
     end
```

Comment:
Replace all the `and` with `&&` because the logical operator is required here.  If subfields are arrays, perhaps ruby built-in == operator will work without the zip (not sure about that).